### ISSUE TITLE
Add optional support for MKL PARDISO solver

### ISSUE THREAD

**Issue by ma-sadeghi** [scipy/scipy#14216]
**Is your feature request related to a problem? Please describe.**
The SciPy's current implementation of `spsolve` is extremely slow. It can be made faster by using `umfpack`, but that's only available for Mac and Linux (in fact, it also supports Windows but up to Python 3.6, so practically it's unsupported).

And even `umfpack` is not so fast. Here's a comparison I made for `umfpack.spsolve` vs. MKL PARDISO for a typical 3d finite difference problem:
<img src="https://user-images.githubusercontent.com/14086031/121534976-b8286380-c9cf-11eb-95a9-13ec949fc680.png" width=50%>
(I'm aware that the speedup really depends on the problem, but the above is just a sample)

Not to mention, for Windows users, the above comparison gets worse by 10X (in my experience, SciPy's default `spsolve` is ~10X slower than `umfpack`'s on average).

**Describe the solution you'd like**
Since most of SciPy users are getting SciPy from `conda` anyway (I might be wrong, I'm just guessing), almost everybody has MKL installed by default. So why not leverage the super fast MKL PARDISO?

The solution I'm proposing (and I'm willing to draft a PR if SciPy devs are OK with the idea) is that similar to the check that SciPy currently does for `umfpack` (i.e., it checks whether or not `umfpack` is available), we introduce another check for MKL, and use PARDISO if MKL is found.

**Describe alternatives you've considered**
None.

**Additional context (e.g. screenshots)**
None.

**Comment by mdhaber** [comment]
Would you propose to introduce an existing wrapper for Pardiso (e.g. [haasad/PyPardiso](https://github.com/haasad/PyPardisoProject)) as an optional dependency, or would you write your own?

**Comment by ma-sadeghi** [comment]
@mdhaber I was thinking of writing a standalone wrapper and include it in SciPy. I already have a working prototype, so it would only be a matter of making it SciPy-compatible, so to speak.

**Comment by mdhaber** [comment]
OK. How would you suggest that the interface be exposed? 
I think it would be nice to add an explicit "solver" argument to `scipy.sparse.linalg.spsolve` and deprecate `use_umfpack`. The default would be `None`, which would maintain the existing behavior - look for `scikit-umfpack`, use it if installed, and fall back to SuperLU otherwise. The `solver` argument would use the specified solver (`{'suplerlu', 'umfpack', 'pardiso'} or error-out if it is not available.
Would Pardiso support `permc_spec`? (Does UMFPACK, do you know?) Would there (also) be a separate function that would expose solver-specific options, or could solver-specific options be included in an `options` structure (like `minimize`)?

**Comment by ma-sadeghi** [comment]
Sorry for the late reply.

I like your `solver` argument very much. Regarding "omnibus" functions, I feel like multiple backends should be fine. I don't think users care too much (if at all) about which solver is being used, as long as we provide the fastest and greatest, and so maybe it's ok to hide them behind the `solver` argument.

I don't know about `lobpcg` (in fact, I didn't know `lobpcg` and `spsolve` were related), but I think that in addition to `spsolve`, we should change `factorized`, so users can reuse the factorization for multiple RHS. So, maybe add the `solver` argument to `factorized` as well?

**Comment by mdhaber** [comment]
> Regarding "omnibus" functions, I feel like multiple backends should be fine.

The reason it gets tricky is because sometimes different backends have very different options. What we try to do is make the common options keyword arguments and the backend-specific options in an `options` dictionary, but this can still be problematic when, say, a new solver is added that doesn't support the old keyword options. And when there is too much divergence, people wonder why there aren't just separate functions.
(That said, I think in this case there is probably enough in common that it is justified to have one function.)

`lobpcg` and `spsolve` aren't really related. I'm just giving an example of a situation that could be similar to this one. There is a generic function for sparse SVD (`scipy.sparse.linalg.svds`), then a second backend got added (LOBPCG) and, in addition to being exposed as a `solver` argument of the generic function, it also got its own function (`lobpcg`).

**Comment by ma-sadeghi** [comment]
> The reason it gets tricky is because sometimes different backends have very different options. What we try to do is make the common options keyword arguments and the backend-specific options in an options dictionary, but this can still be problematic when, say, a new solver is added that doesn't support the old keyword options. And when there is too much divergence, people wonder why there aren't just separate functions.
(That said, I think in this case there is probably enough in common that it is justified to have one function.)

Fair enough.

> lobpcg and spsolve aren't really related. I'm just giving an example of a situation that could be similar to this one. There is a generic function for sparse SVD (scipy.sparse.linalg.svds), then a second backend got added (LOBPCG) and, in addition to being exposed as a solver argument of the generic function, it also got its own function (lobpcg).

Oh, I see! It makes sense. If I'm not mistaken, SciPy currently exposes `scipy.sparse.linalg.SuperLU`, we could expose a similar object called `scipy.sparse.linalg.Pardiso`, which would then be used in `spsolve` as a wrapper. Does that make sense?

**Comment by mdhaber** [comment]
>  If I'm not mistaken, SciPy currently exposes ... Does that make sense?

Yes, it makes sense. One thing that is a little different is that `SuperLU` objects are not instantiated by the user; they are returned by `splu` and `spilu`. This leads to the question "How would `Pardiso` objects be instantiated?"

(And considering that we are also thinking about adding better support for UMFPACK and CHOLMOD, how do these fit into the puzzle?)

I'll be thinking about this some more, too. 

UI will take some time to work out, but that doesn't have to stop us from taking a look at your wrapper. I'd be interested in taking a look. One way to do this is to create a draft PR here on SciPy's repo, but another thing we do sometimes is create a PR on a private fork of SciPy, get it mostly ready there, then submit a PR to the main repo when it's more or less the way we want it. (For example, @mckib2 and I are currently working on a wrapper for PROPACK over at mckib2/scipy#9. When we're happy with it, we'll open a PR here.) If you set up a PR on your own fork for this, let me know and I'll take a look at the diff and try it out.

**Comment by ngodber** [comment]
A related topic is https://github.com/scipy/scipy/issues/14096  One of the big speed improvements that Pardiso brings is that it is multithreaded whereas (non ufmpack) are not. @ma-sadeghi there are some wrappers available which are quite robust https://github.com/simpeg/pymatsolver https://github.com/dwfmarchant/pyMKL

**Comment by ngodber** [comment]
Benefit of having a fast non-MKL solution is it avoids MKL shenanigans - https://www.pugetsystems.com/labs/hpc/How-To-Use-MKL-with-AMD-Ryzen-and-Threadripper-CPU-s-Effectively-for-Python-Numpy-And-Other-Applications-1637/
